csharp_dotnetcore/60.ConversationHistory/ConversationHistoryBot.cs

```diff
@@ -0,0 +1,89 @@
+    public class ConversationHistoryBot : IBot
+    {
+        private readonly TranscriptStore _transcriptStore;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ConversationHistoryBot"/> class.
+        /// </summary>
+        /// <param name="transcriptStore">A class containing <see cref="IStatePropertyAccessor{T}"/> used to manage state.</param>
+        public ConversationHistoryBot(TranscriptStore transcriptStore)
+        {
+            _transcriptStore = transcriptStore;
+        }
+
+        /// <summary>
+        /// Every conversation turn for our Echo Bot will call this method.
+        /// There are no dialogs used, since it's "single turn" processing, meaning a single
+        /// request and response.
+        /// </summary>
+        /// <param name="turnContext">A <see cref="ITurnContext"/> containing all the data needed
+        /// for processing this conversation turn. </param>
+        /// <param name="cancellationToken">(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A <see cref="Task"/> that represents the work queued to execute.</returns>
+        /// <seealso cref="BotStateSet"/>
+        /// <seealso cref="ConversationState"/>
+        /// <seealso cref="IMiddleware"/>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (turnContext.Activity.Type == ActivityTypes.Message)
+            {
+                // Echo back to the user whatever they typed.
+                var responseMessage = $"You sent '{turnContext.Activity.Text}'\n";
+                await turnContext.SendActivityAsync(responseMessage);
+            }
+
+            // Send over the transcript when a request arrives. This is something that the client and the Bot have
+            // to agree on.
+            var shouldUploadHistory = false;
+            if (shouldUploadHistory)
+            {
+                var connectorClient = turnContext.TurnState.Get<ConnectorClient>(typeof(IConnectorClient).FullName);
+                var activity = turnContext.Activity;
+
+                // Get all the message type activities from the Transcript.
+                string continuationToken = null;
+                var activities = new List<Activity>();
+                do
+                {
+                    var pagedTranscript = await _transcriptStore.Store.GetTranscriptActivitiesAsync(activity.ChannelId, activity.Conversation.Id);
+                    activities.AddRange(pagedTranscript.Items
+                                        .Where(a => a.Type == ActivityTypes.Message)
+                                        .Select(ia => (Activity)ia)
+                                        .ToList());
+                    continuationToken = pagedTranscript.ContinuationToken;
```

Comment:
In terms of LINQ feedback, you should be able to avoid the ToList() here and just do one at the end.

Reply:
Not applicable anymore since I am calling upload inside the loop.

Reply:
Not applicable anymore since I am calling upload inside the loop.

---
In reply to: [218234175](https://github.com/Microsoft/BotBuilder-Samples/pull/391#discussion_r218234175) [](ancestors = 218234175)